compose: Add wildcard mentions check on client side.

shameondev opened the pull request:
This PR is intended to implement client-side checking of wildcard mentions in message edit mode, while preserving the existing behavior from the compose box. The main changes made to accomplish the goal is:
1. Added a `$banner_container` parameter to certain functions, such as `show_wildcard_warnings` in the `compose_validate` module. This allows for their reuse in both the compose box and edit context.
2. Significantly changed the way wildcard banners are tracked. The system has become more complex, as we now need to track banners not only for the compose box but also for multiple edit inputs. The system now checks if the banner has been displayed for each input. If the banner has already been shown, we skip displaying it, return true for validation, and proceed with the API request.
3. Added message mention validation for message editing. The behavior will be similar to the validation found in the compose box when a user sends a message.

Fixes: #25411.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![2023-05-05 at 18 53 26 - Amethyst Fowl](https://user-images.githubusercontent.com/53193850/236451200-b10d0594-d761-45c8-a56d-f82dab5b70d5.gif)

timabbott commented on web/src/compose.js:
```diff
@@ -471,9 +471,17 @@ export function initialize() {
+            } else {
+                finish();
+            }
+            compose_validate.remove_wildcard_banner_shown_for_container($banner_container);
```
Hmm, is this the right time to remove this? I'd expect us to want to do it in the `success` handler for the message actually having been sent / saved when the compose box is cleared if locally echoed.

shameondev replied:
Yeah, the timing wasn't right, so I moved it to those success handlers.

timabbott commented on web/src/compose_validate.js:
```diff
@@ -355,9 +356,18 @@ export function set_wildcard_mention_large_stream_threshold(value) {
+        // We will not update wildcard mention if we came from compose_typeahead
+        // We are sure that wildcard_mention exists, so we don't need to check it here.
+        wildcard_mention = "";
+    }
```
What would happen if we didn't do this check?

timabbott replied:
Also, we're not setting this value to `null`, so I'm confused -- won't the below check still run?

shameondev replied:
In the first iteration, it helped to control the visibility of the banner. It was definitely a bug somewhere else, but I was able to fix it this way. Now this logic has been removed.

timabbott commented on web/src/composebox_typeahead.js:
```diff
@@ -1016,6 +1032,10 @@ export function initialize_topic_edit_typeahead(form_field, stream_name, dropup)
 }
 
+function is_wildcard_mention(mention) {
+    return ["all", "everyone", "stream"].some((wildcard) => mention.includes(wildcard));
```
Why is `includes` the right check -- don't we need to check for equality? Will someone with the name `Ballin" trigger this warning?

timabbott replied:
There's a function `full_name_matches_wildcard_mention` in `people.js`; you can do a prep commit to rename that to `is_wildcard_mention` and export it, and then just call that here.

shameondev replied:
> Why is includes the right check -- don't we need to check for equality? Will someone with the name `Ballin" trigger this warning?

Yeah, you're right. This could result in a false positive match. I've made some changes to `full_name_matches_wildcard_mention` to ensure that it checks whether the mention contains a wildcard symbol.

shameondev replied:
I'm considering where we can place the `is_wildcard_mention` function. It seems like a utility function that is not related to either `people.js` or `composebox_typeahead.js`. Is there a designated location for functions like this?

timabbott replied:
`people.js` is probably fine for this for now.

timabbott commented on web/src/composebox_typeahead.js:
```diff
@@ -820,6 +821,21 @@ export function content_typeahead_selected(item, event) {
+                const edit_stream_id = $textbox
+                    .closest(".recipient_row")
+                    .find(".message_header_stream")
+                    .attr("data-stream-id");
```
What happens if you're editing a private message?

shameondev replied:
Yeah, there will be an error since we don't have any information about the stream.

Currently, we have removed this logic because we don't want to check for wildcard mentions before we click "save" or "edit".

timabbott commented on web/src/composebox_typeahead.js:
```diff
@@ -820,6 +821,21 @@ export function content_typeahead_selected(item, event) {
+                    .closest(".recipient_row")
+                    .find(".message_header_stream")
+                    .attr("data-stream-id");
+                const stream_id = Number(compose_stream_id || edit_stream_id);
```
This logic doesn't make any sense. If we're typing in the compose box, we need to use the compose box's stream ID. Otherwise we use the edit box stream ID. Under no circumstances does it make sense to use the compose box stream ID if we're editing a message; that is introducing another instance of #25410.

shameondev replied:
Got it! Thanks so much for the explanation.

timabbott commented on web/src/compose_validate.js:
```diff
@@ -368,27 +378,27 @@ function validate_stream_message_mentions(stream_id) {
+                $banner_container.closest("form").attr("id"),
+            )
+        ) {
             // user has not seen a warning message yet if undefined
```
This comment is now wrong, I think?

It sounds like the logic here, though, is that we're now keeping track of whether we've shown the banner for this input in this session, rather than whether the user has acknowledged it; is the intention that if the user hides the banner by clicking the close button, we won't clear the state that we've shown the banner, and that's what we're achieving? I think that's correct, but want to make sure I understand correctly.

shameondev replied:
Yeah, the comment is now wrong. 

> we're now keeping track of whether we've shown the banner for this input in this session, rather than whether the user has acknowledged it

that is right.

> is the intention that if the user hides the banner by clicking the close button, we won't clear the state that we've shown the banner, and that's what we're achieving?

Currently, this state is just keeping track of which banners are being shown to the user. While I was thinking about your question, I wondered if it was really necessary. If it's only doing such a simple job of tracking, there might be easier options to avoid adding multiple banners. Perhaps it's worth taking another look at `is_wildcard_banner_shown_for_container` in this iteration.

shameondev replied:
I have given this more thought, and I believe you are correct. Currently, we are collecting the IDs of the inputs for which we have displayed the warning banner. We only remove the ID from the set after the message has been sent or saved.

timabbott commented on web/src/compose_validate.js:
```diff
@@ -230,38 +230,39 @@ export function warn_if_topic_resolved(topic_changed) {
-            $("#compose_banners"),
-        );
+    if ($banner_container.find(`.${CSS.escape(classname)}`).length === 0) {
+        compose_banner.append_compose_banner_to_banner_list(wildcard_template, $banner_container);
```
We likely need to do something so the button says "Yes, save" when editing.

shameondev replied:
I think it would be appropriate, since clicking on the button saves the message, even though in the end we send it and use the "send" word in the main message.
![image](https://user-images.githubusercontent.com/53193850/236433834-c27efbd9-e21e-4085-87a2-73242d7f81d1.png)

timabbott replied:
Sounds good.

shameondev commented:
Thank you for the review, @timabbott. I believe I may have misunderstood the problem we are trying to solve, and ended up focusing on the wrong issue. What would you suggest we do with this PR? For now, I think I will open a new one.

shameondev commented:
I may have overreacted by saying that we need a new PR. There are a lot of similarities between the current version and the one we should end up with. So, I will work on fixing this instead of creating a new PR.

shameondev commented:
I have updated the PR, taking into account all the necessary details i missed in last iteration. Additionally, I have conducted more thorough manual testing. 

Currently, I have not fixed the node tests despite dedicating some time to it. I will start a topic at CZO to seek help.

timabbott commented on web/src/people.js:
```diff
@@ -1198,8 +1198,18 @@ export function get_mention_syntax(full_name, user_id, silent) {
+export function is_wildcard_mention(mention) {
+    const name = extract_name_from_mention(mention);
+    return ["all", "everyone", "stream"].includes(name);
 }
```
Do callers know whether they're expecting to have a name or mention syntax? I would usually expect so, and that's the case for `get_mention_syntax` above.

For that reason I think it'd be better to have two functions; `is_wildcard_mention` and `is_wildcard_mention_syntax`, with the latter requiring something starting with `@`, and the former requiring it not to.

shameondev replied:
The callers are aware of what they're passing, yeah.

I was wondering, do you think it would be a good idea to raise a blueslip error or a warning if someone accidentally passes something starting with '@' to the wrong function? I'm not sure how often this will come up, but it might be worth considering just to be on the safe side. What do you think?

shameondev replied:
This changes was rolled back.

timabbott commented on web/tests/people.test.js:
```diff
@@ -1077,6 +1077,15 @@ test_people("get_mention_syntax", () => {
+    assert.ok(people.is_wildcard_mention("@_**all**"));
+    assert.ok(people.is_wildcard_mention("@**everyone**"));
+    assert.ok(people.is_wildcard_mention("@**stream**"));
+})
```
This test is problematic, in that it doesn't exercise the case without an `@` at all.

timabbott replied:
OK.

timabbott commented:
I posted one comment on the first commit; haven't read the second, but it'd be very helpful if you split it into smaller commits and posted something about it's technical design on Monday so I can understand what I'm trying to verify.

shameondev commented:
Hey @timabbott! Just wanted to let you know that I took your advice and split up the work into smaller commits. I also added more context to the PR description, although it's similar to the commit history and commit messages. I'm not sure if it's enough for a "technical design," but feel free to let me know if you need more information!

timabbott commented on web/tests/lib/compose_banner.js:
```diff
@@ -12,4 +12,27 @@ exports.mock_banners = () => {
     $("#compose_banners .warning").remove = () => {};
     $("#compose_banners .error").remove = () => {};
+
+    const $stub = $.create("stub_to_remove");
```
I'm not sure; @evykassirer may have thoughts as the original author of this test file.

timabbott commented on web/src/message_edit.js:
```diff
@@ -855,6 +856,13 @@ export function save_inline_topic_edit($row) {
+    const stream_id = Number.parseInt(
+        rows.get_message_recipient_header($row).attr("data-stream-id"),
+        10,
+    );
```
I'm concerned that this will throw an exception when editing a private message.

shameondev replied:
If the attribute `data-stream-id` does not exist, calling `attr("data-stream-id")` will return `undefined`. Subsequently, passing `undefined` to `parseInt` will result in the return of `NaN`. This approach looks viable.

timabbott commented on web/src/compose_validate.js:
```diff
@@ -251,18 +251,25 @@ function show_wildcard_warnings(stream_id, $banner_container, wildcard_mention)
+export function delete_wildcard_banner_shown_for_container($banner_container) {
+    shown_wildcard_banners_for_container.delete(
+        $banner_container.closest("form").attr("id") ?? "send_message_form",
+    );
```
Maybe this ends up being fine because the forms will have similar roles, what happens if you start editing a message in the "all messages" view and then navigate to another view? I guess since the form ID essentially encodes the message ID, it may reuse the container ID as well ... but that might not be wrong since it'll always be a different message edit form for the same message?

shameondev replied:
> it'll always be a different message edit form for the same message?

Yeah, that is right!

timabbott commented on web/src/compose_validate.js:
```diff
@@ -251,18 +251,25 @@ function show_wildcard_warnings(stream_id, $banner_container, wildcard_mention)
-    user_acknowledged_wildcard = value;
+export function add_wildcard_banner_shown_for_container($banner_container) {
+    shown_wildcard_banners_for_container.add(
+        $banner_container.closest("form").attr("id") ?? "send_message_form",
```
This calculation would be much better done as `get_wildcard_banner_key_for_container($banner_container)` to avoid duplicating it several times.

shameondev replied:
Sure, thanks!

timabbott commented on web/src/compose.js:
```diff
@@ -482,7 +483,6 @@ export function initialize() {
             const {$banner_container} = get_input_info(event);
             compose_validate.clear_wildcard_warnings($banner_container);
-            compose_validate.set_user_acknowledged_wildcard_flag(true);
             finish();
```
So, this, as written, makes clicking the usual "send" button send/save the message, rather than the previous behavior of requiring you to actually click the "Yes, send" version implemented here.

I believe this is easy to fix without the previous complexity of that global variable: just push a parameter `{acknowledged_wildcard_mention_warning: true}` here and plumb it through.

That said, if you're going to do that, I'm not sure I understand the function of methods like `add_wildcard_banner_shown_for_container` -- ultimately we just need to know which button was pushed to decide whether to actually send the message or not.

timabbott commented:
The commit structure is indeed much easier to read, and that helped a lot in my being able to find the issues noted above, so thanks for doing that. But those issues need to be addressed before we'll be able to integrate this.

Additionally, please play around with this as self-QA a bit more once you've addressed the feedback (which may end up fixing the bugs as a side effect) -- I was able to in a minute of fiddling get into a state in the compose box where the banner didn't appear when trying a send a new message -- it just failed silently. I don't know precisely what I did, but I expect you should be able to trigger it.

shameondev commented:
> I was able to in a minute of fiddling get into a state in the compose box where the banner didn't appear when trying a send a new message -- it just failed silently. I don't know precisely what I did, but I expect you should be able to trigger it.

Hmm, I've been attempting to reproduce the issue but have been unsuccessful thus far. I suspect this may be due to the recent rollback of the changes related to tracking which banners are displayed. This simplification of the system should result in fewer bugs and fewer corner cases that you may have encountered. Please let me know if you continue to experience any unusual behavior.

alya commented:
Thanks! If the message already includes a wildcard mention, we should not show the banner when it's edited. It's not necessary, and the banner message is misleading, as the user is not sending any notifications at that time (unless stream membership has changed).

alya commented:
Looks good to me in manual testing other than that.

timabbott commented:
This seems workable; it's still a logic bug that we share the global variable `user_acknowledged_wildcard`, rather than just passing the value through, but it seems unlikely to come up in practice.

Merging this since it's considerably better than main.

I opened as a follow-up https://github.com/zulip/zulip/issues/25507 to track that we need to fix the bug this introduces with editing wildcard mention messages.

timabbott commented:
Well I suppose it was a bit silly to open #25507, given that I fixed it before CI finished, but in any case, I've added a couple commits to address that. 

Thanks for all the work on this @shameondev!

shameondev commented:
Thank you for the suggestion and for helping me with this PR and addressing the issue of editing the message that includes a wildcard mention!

I fixed the failing test by adding `wildcard_mention: false` to the hardcoded object that we use to assert the result. I have not disabled the auto-merge since it's something trivial, but I just wanted to let you know, @timabbott.
